Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py

```diff
@@ -22,60 +22,86 @@
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 import os
+import re
 import sys
 import time
 
+from webkitcorepy import string_utils
+
 if sys.platform.startswith('win'):
     import msvcrt
-else:
-    import fcntl
 
 
 class FileLock(object):
+    INTEGER_RE = re.compile(r'^\d+$')
+    USE_WINDOWS = sys.platform.startswith('win')
+    USE_EXLOCK = not USE_WINDOWS and getattr(os, 'O_EXLOCK', False)
+
+    @classmethod
+    def is_process_running(cls, pid):
+        try:
+            os.kill(pid, 0)
+            return True
+        except OSError:
+            return False
 
     def __init__(self, path, timeout=20):
         self.path = path
         self.timeout = timeout
         self._descriptor = None
+        self._depth = 0
 
     @property
     def acquired(self):
         return bool(self._descriptor)
 
     def acquire(self):
         if self._descriptor:
+            self._depth += 1
             return True
 
-        descriptor = os.open(self.path, os.O_TRUNC | os.O_CREAT)
+        if not self.USE_EXLOCK and not self.USE_WINDOWS and os.path.exists(self.path):
+            with open(self.path) as file:
+                pid = file.readline().strip()
+            if self.INTEGER_RE.match(pid) and not self.is_process_running(int(pid)):
+                os.unlink(self.path)
+
         start_time = time.time()
         while True:
             try:
-                if sys.platform.startswith('win'):
+                if self.USE_WINDOWS:
+                    self._descriptor = os.open(self.path, os.O_TRUNC | os.O_CREAT)
                     msvcrt.locking(descriptor, msvcrt.LK_NBLCK, 32)
+                elif self.USE_EXLOCK:
+                    self._descriptor = os.open(self.path, os.O_CREAT | os.O_WRONLY | os.O_EXLOCK | os.O_NONBLOCK)
                 else:
-                    fcntl.flock(descriptor, fcntl.LOCK_EX | fcntl.LOCK_NB)
-                self._descriptor = descriptor
+                    self._descriptor = os.open(self.path, os.O_CREAT | os.O_WRONLY | os.O_EXCL)
+                    os.write(self._descriptor, string_utils.encode(str(os.getpid())))
+                self._depth += 1
                 return True
 
-            except IOError:
+            except (IOError, OSError):
                 if time.time() - start_time > self.timeout:
-                    os.close(descriptor)
+                    if self._descriptor:
+                        os.close(self._descriptor)
                     self._descriptor = None
                     return False
                 time.sleep(0.01)
```

Comment:
Not very happy about spinning, but I guess there is no other way if we want to have a timeout for locking. Maybe we don't have to, though?

Reply:
There _is_ another way for `O_EXLOCK` (although it doesn't work for Windows and ports without `O_EXLOCK`, so we have to keep the spinning). Updating PR with the approach that doesn't use a spin lock.